haystack/preview/components/samplers/top_p.py

```diff
@@ -0,0 +1,131 @@
+import logging
+from typing import List, Optional, Dict, Any
+
+from haystack.preview import ComponentError, Document, component, default_from_dict, default_to_dict
+from haystack.preview.lazy_imports import LazyImport
+
+logger = logging.getLogger(__name__)
+
+
+with LazyImport(message="Run 'pip install torch>=1.13'") as torch_import:
+    import torch
+
+
+@component
+class TopPSampler:
+    """
+    Selects documents based on the cumulative probability of the Document similarity scores using top p (nucleus)
+    sampling.
+
+    Usage example:
+
+    ```python
+    from haystack.preview import Document
+    from haystack.preview.components.samplers import TopPSampler
+
+    sampler = TopPSampler(top_p=0.95)
+    docs = [
+        Document(text="Berlin", metadata={"similarity_score": -10.6}),
+        Document(text="Belgrade", metadata={"similarity_score": -8.9}),
+        Document(text="Sarajevo", metadata={"similarity_score": -4.6}),
```

Comment:
It feels kind of odd to have the similarity scores be negative? Typically these scores will be positive, but good to see that this works for both!

But for the example, maybe having positive scores would be better to see. What do you think?

Reply:
I took values from SimilarityRanker and those are negative, these are essentially verbatim from that component. Let's test both

Reply:
Ahh okay I see, that's typically because in V1 we would put the raw logits through a Sigmoid layer before being stored as scores.

Reply:
Let's add that in the second pass over SimilarityRanker and then a good test would be no regressions. Marking as resolved